File: code/python/DNA.py
class DNASolution:
    def __init__(self):
        self.dna = None
        self.dnaBases = {'A': 0, 'C': 0, 'G': 0, 'T': 0}

        self.readDnaSeq()
        self.countDnaBases()

    # lendo a sequência do dna
    def readDnaSeq(self):
        with open('C:\Programming\Rosalind\input/rosalind_dna.txt', 'r') as input_data:
            self.dna = input_data.read()
            input_data.close()

    # contando a qtd. de cada base (A, C, G, T)
    def countDnaBases(self):
        for base in self.dna:
            if base in self.dnaBases:
                self.dnaBases[base] += 1

    # resultado da qtd. de cada base
    def numberDnaBases(self):
        bases = list(self.dnaBases.values())

        with open('C:\Programming\Rosalind\output\DNA.txt', 'w') as output_data:
            for amount in bases:
                output_data.write(f"{amount: <5}")
            output_data.close()


if __name__ == "__main__":
    DNASolution().numberDnaBases()

File: code/c++/DNA.cpp
#include <iostream>
#include <map>
#include <fstream>
#include <string>

class DNASolution
{
private:
    std::string dna_seq;
    std::map<char, int> dna_bases = {
        {'A', 0}, // Adenina
        {'T', 0}, // Timina
        {'C', 0}, // Citocina
        {'G', 0}  // Guanina
    };

public:
    // Lendo a sequência do DNA
    void read_dna_seq()
    {
        std::ifstream input_data("C:/Programming/Rosalind/input/rosalind_dna.txt");

        if (input_data.is_open())
        {
            // lendo os dados
            std::getline(input_data, dna_seq, '\n');

            input_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }

    // Contando a qtd. de cada base do DNA
    void count_dna_bases()
    {
        for (auto base : dna_seq)
        {
            dna_bases[base]++;
        }
    }

    // Retornando o número de cada base do DNA
    void number_of_each_dna_bases()
    {
        read_dna_seq();
        count_dna_bases();

        std::ofstream output_data("C:/Programming/Rosalind/output/DNA.txt");

        if (output_data.is_open())
        {
            for (auto base : dna_bases)
            {
                output_data << base.second << " ";
            }
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }
};

int main()
{
    DNASolution solution;
    solution.number_of_each_dna_bases();

    return 0;
}